Add company timeline view

reupen opened the pull request:
Issue number: N/A

### Description of change

This adds a view to retrieve the timeline for a company using the reporting service.

A new route has been added at `/v3/company/<id>/timeline`. A new permission `company.read_company_timeline` has been also added.

### Checklist

* [x] Have any relevant search models been updated?
* [x] Have any relevant fixtures (`fixtures/test_data.yaml`) been updated?
* [x] Have any relevant select-/prefetch-related field lists in the views and search apps been updated?
* [x] Has the admin site been updated (for new models, fields etc.)?
* [x] Has the README been updated (if needed)?

marcofucci commented on datahub/company/timeline/views.py:
```diff
@@ -0,0 +1,40 @@
+
+    def list(self, request, *args, **kwargs):
+        """Lists timeline events (paginated)."""
+        company = self.get_object()
```
This (`self.get_object`, `queryset` etc.) works but it's a bit weird as it refers to a company model whilst this ViewSet is for timeline events instead. Not a big deal I suppose.

reupen replied:
Yeah, it could be a method on CompanyViewSet in theory, but I went for keeping all the timeline stuff separate so that it's mostly self-contained. I agree that it's a little bit odd, but it does save us from duplicating that logic.

marcofucci commented on datahub/company/timeline/client.py:
```diff
@@ -0,0 +1,80 @@
+        if api_id is None:
+            api_id = settings.REPORTING_SERVICE_API_ID
+        if api_key is None:
+            api_key = settings.REPORTING_SERVICE_API_KEY
```
I'm not sure `api_url=None, api_id=None, api_key=None` adds anything here, I suspect reading from settings should be enough. In tests, we can override settings to test stuff.

reupen replied:
Have removed these now.

marcofucci commented on datahub/company/test/timeline/test_client.py:
```diff
@@ -0,0 +1,128 @@
+    '/api/v1/company/events/?companies_house_id=125695': {
+        'status_code': status.HTTP_404_NOT_FOUND
+    },
+    '/api/v1/company/events/?companies_house_id=125696': {
```
A small thing but using a company number very different from `125695` would make scanning the code easier as it would make it obvious that the numbers are different.

reupen replied:
I've updated these.

marcofucci commented on datahub/company/test/timeline/test_views.py:
```diff
@@ -0,0 +1,116 @@
+        url = reverse('api-v3:company:timeline-collection', kwargs={'pk': company.pk})
+
+        response = api_client.get(url)
+        assert response.status_code == status.HTTP_403_FORBIDDEN
```
I would add another couple of tests:

1. when the response from the API is different from what we expect and the serializer.is_valid fails
2. when the response form the API returns a 500 error.

elcct replied:
I agree with above.

reupen replied:
I've added a couple more tests for those cases.

reupen commented:
> I've added a couple of comments but I'm also worried that the name reporting service misleads a bit. Why don't we call it timeline service or something similar?

I started off calling it company timeline API, but I changed it because:

- the service is actually called something like data platform reporting service
- it does more than just company timeline

See https://github.com/uktrade/dt07-reporting/blob/develop/api.md.

So calling it company timeline API is wrong and won't make sense if we start using anything else from the service. Any other suggestions for names?

reupen commented:
I've renamed the settings and classes so that they refer to the 'data science company API' instead.
